### PR TITLE
Testing with pytest

### PR DESCRIPTION
**Pull request by vasthecat**

### CODE REVIEW

File: tests/test_multithreaded_iterator.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    return sb.run(command.split(), capture_output=True) \
+               .stdout \
+               .decode() \
+               .strip()
+
+def hash_password(password, salt="hi"):
+    return run(f"./encr -p {password} -s {salt}")
+
+# Length 1
+def test_st_iter_length1_first():
```
Comment (hnarkaytis) [inline tests/test_multithreaded_iterator.py]: Too many copy/paste. Repeated code should be factored out to a function. Majority of the tests will be a call of this function

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+on: [push, pull_request]
+jobs:
+  test-project:
+    runs-on: ['ubuntu-latest']
```
Comment (hnarkaytis) [inline .github/workflows/tests.yml:4]: What about macos?

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml (hunk 2 above)
Comment (vasthecat) [reply .github/workflows/tests.yml:4]: Github seems to have problems finding macos runners. Also macos has some strange bug when program halts for no reason in multithreaded/generator mode. I'd say that fixing those problems is not in priority at that moment.

File: tests/test_generator_iter.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    assert result == f"Password found: '{password}'"
+
+
+def base_call_notfound(password, alphabet="abc"):
+    length = len(password)
+    hashed = hash_password(password, "hi")
+    result = run(f"./brute -g -i -l {length} -h {hashed} -a {alphabet}")
+    assert result == f"Password not found"
+
+# Length 1
```
Comment (hnarkaytis) [inline tests/test_generator_iter.py]: Standard practice is to have:
1. One happy path test: "abc", not found
2. Tests for all corner cases: "aaa", "ccc"
3. Performance test: adaptive test that will find password length for which run takes some measurable time (>0.1s) and will check that for length+1 consumed time grew linearly (on a length of alphabet).

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml (hunk 2 above)
Comment (hnarkaytis) [reply .github/workflows/tests.yml:4]: probably they limit number of created threads. You should add this as a command line parameter and test.

File: tests/test_generator_iter.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    base_call_found("alliedd", "adefil")
+
+def test_gen_iter_bigger_notfound():
+    base_call_notfound("hellodf", "defghl")
+
+# Performance test
+from time import time
+def test_performance():
+    t1 = time()
+    base_call_notfound("qqqqqqqqqq")
```
Comment (hnarkaytis) [inline tests/test_generator_iter.py]: with 3 chars alphabet I would expect run time for this to be below measurable threshold. You need to dynamically find required password length, but not to hard code into the test.

File: tests/test_generator_iter.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+def test_gen_iter_length1():
+    base_call_found("b")
+
+def test_gen_iter_length1_last():
+    base_call_found("c")
+
+def test_gen_iter_length1_notfound():
+    base_call_notfound("q")
+
+# Length 2
```
Comment (hnarkaytis) [inline tests/test_generator_iter.py]: Length 1 and 7 should be enough. No need for 2 & 3

File: tests/test_generator_iter.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
+import subprocess as sb
+
+
+def run(command):
```
Comment (hnarkaytis) [inline tests/test_generator_iter.py]: It make sense to deduplicate common code between test into shared library

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

File: tests/test_generator_iterator.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+# Bigger alphabet
+def test_gen_iterator_bigger():
+    base_call_found("alliedd", "adefil")
+
+def test_gen_iterator_bigger_notfound():
+    base_call_notfound("hellodf", "defghl")
+
+# Performance test
+from time import time
+def test_performance():
```
Comment (hnarkaytis) [inline tests/test_generator_iterator.py]: Why not to use method from runners.py?

Comment (hnarkaytis) [inline tests/test_generator_iterator.py]: Why not to use method from runners.py?

**Review by hnarkaytis — CHANGES_REQUESTED** [review]
It looks like tests for different run modes are absolutely identical and support of copy/paste over multiple files takes too much effort. I believe that all tests should be the same and you need only to iterate over different run modes, i.e. set of command line flags.

**Review by hnarkaytis — APPROVED** [review]
Much better
